### DISCUSSION TITLE
Error: The edge runtime does not support Node.js 'crypto' module

### DISCUSSION THREAD

**Discussion by amirhhashemi** [vercel/next.js#51753] [Help]
### Summary

I'm trying to implement a strict CSP with nonce and I use the `crypto` module to generate the nonce in a middleware. But when I run the app in dev mode I get this error in my terminal:

```tsx
- error The edge runtime does not support Node.js 'crypto' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
> 1 | import crypto from "crypto";
  2 | import { env } from "@/app/env.mjs";
  3 | const IS_PRODUCTION = env.NODE_ENV === "production";
  4 | export function getCSP() {
```

The problem is I don't use Edge runtime. I started a brand new Next.js app with `create-next-app` today. I haven't changed any next.js configuration. I believe the default runtime is `nodejs` so I shouldn't get that error. 

This is my `next.config.js` file:
```
/** @type {import('next').NextConfig} */
const nextConfig = {};

export default nextConfig;
```
Why I get that error and how can I fix it?

### Additional information

```js
Node.js version: `16.20.0`
Next.js version: `13.4.7`
Package manager `pnpm v8.2.0`
```

**Comment by amirhhashemi** [comment]
I found that middlewares use the Edge runtime by default and there is no way to change that currently.

https://github.com/vercel/next.js/discussions/34179

**Reply by gerardnico** [reply]
What a shame and what is funny is that the [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/router-handlers) supports NodeJs and crypto ???.

**Reply by theDanielJLewis** [reply]
So what's the solution, if there is one?

**Reply by amirhhashemi** [reply]
Haven't found a solution, but you can move your logic to an API endpoint and call it from the middleware.

Or if it's possible use an alternative method that doesn't require Node.js. In my case, I just needed a random string so I used `nanoid` which is not dependent on Node.js.
